once_test.py

```diff
@@ -214,27 +432,25 @@ def test_threaded_multiple_functions(self):
     def test_different_fn_do_not_deadlock(self):
         """Ensure different functions use different locks to avoid deadlock."""
 
-        fn2_called = False
+        fn1_called = threading.Event()
+        fn2_called = threading.Event()
 
         # If fn1 is called first, these functions will deadlock unless they can
         # run in parallel.
         @once.once
         def fn1():
-            nonlocal fn2_called
-            start = time.time()
-            while not fn2_called:
-                if time.time() - start > 5:
-                    self.fail("Function fn1 deadlocked for 5 seconds.")
-                time.sleep(0.01)
+            fn1_called.set()
+            if not fn2_called.wait(5.0):
+                self.fail("Function fn1 deadlocked for 5 seconds.")
 
         @once.once
         def fn2():
-            nonlocal fn2_called
-            fn2_called = True
+            fn2_called.set()
 
-        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
-            fn1_promise = executor.submit(fn1)
-            executor.submit(fn2)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
+            fn1_promise = ensure_started(executor, fn1)
```

Comment:
I'm not sure you need ensure_started since you wait the event on 452 anyway. I think that's actually probably true for all usages of `ensure_started`. Since it just sets an event at the beginning of the thread running, every test that wants to verify particular concurrency needs another concurrency primitive within the onced function itself, making `ensure_started` mostly redundant.

Reply:
This is needed in `test_iterator_lock_not_held_during_evaluation`, where we need to ensure that a job has both started and not finished. I added a comment there. Generally you are correct though.

Reply:
Even in that one, we don't need to ensure that the job has started, rather that the once code has been entered (which is done via counter.ready in that test).